cmd/pebble/cmd_logs.go

```diff
@@ -0,0 +1,146 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/signal"
+	"strconv"
+	"time"
+
+	"github.com/jessevdk/go-flags"
+
+	"github.com/canonical/pebble/client"
+)
+
+const (
+	logTimeFormat = "2006-01-02T15:04:05.000Z07:00"
+)
+
+type cmdLogs struct {
+	clientMixin
+	Follow     bool   `short:"f" long:"follow"`
+	Output     string `short:"o" long:"output"`
+	NumLogs    string `short:"n" long:"number"`
+	Positional struct {
+		Services []string `positional-arg-name:"<service>"`
+	} `positional-args:"yes"`
+}
+
+var logsDescs = map[string]string{
+	"follow": "Follow (tail) logs for given services until Ctrl-C pressed.",
+	"output": "Output format: \"text\" (default), \"json\" (JSON lines), or \n\"raw\" (copy raw log bytes to stdout and stderr).",
+	"number": "Number of logs to show (before following); defaults to 10.\nIf 'all', show all buffered logs.",
+}
+
+var shortLogsHelp = "Fetch service logs"
+var longLogsHelp = `
+The logs command fetches buffered logs from the given services (or all services
+if none are specified) and displays them in chronological order.
+`
+
+func (cmd *cmdLogs) Execute(args []string) error {
+	var numLogs int
+	switch cmd.NumLogs {
+	case "":
+		numLogs = 10
+	case "all":
+		numLogs = -1
+	default:
+		var err error
+		numLogs, err = strconv.Atoi(cmd.NumLogs)
+		if err != nil || numLogs < 0 {
+			return fmt.Errorf(`expected n to be a non-negative integer or "all", not %q`, cmd.NumLogs)
+		}
+	}
+
+	var writeLog client.WriteLogFunc
+	switch cmd.Output {
+	case "", "text":
+		writeLog = writeLogText
+	case "json":
+		writeLog = writeLogJSON
+	case "raw":
+		writeLog = writeLogRaw
+	default:
+		return fmt.Errorf(`invalid output format (expected "json", "text", or "raw", not %q)`, cmd.Output)
+	}
+
+	opts := client.LogsOptions{
+		WriteLog: writeLog,
+		Services: cmd.Positional.Services,
+		NumLogs:  &numLogs,
+	}
+	var err error
+	if cmd.Follow {
+		// Stop following when Ctrl-C pressed (SIGINT).
+		ctx := notifyContext(context.Background(), os.Interrupt)
+		err = cmd.client.FollowLogs(ctx, &opts)
+	} else {
+		err = cmd.client.Logs(&opts)
+	}
+	return err
+}
+
+// Needed because signal.NotifyContext is Go 1.16+
+func notifyContext(parent context.Context, signals ...os.Signal) context.Context {
+	ctx, cancel := context.WithCancel(parent)
+	ch := make(chan os.Signal)
+	signal.Notify(ch, signals...)
+	go func() {
+		// Wait for signal, then cancel the context.
+		<-ch
+		cancel()
+	}()
+	return ctx
+}
+
+func writeLogText(timestamp time.Time, service, message string) error {
+	suffix := ""
+	if len(message) == 0 || message[len(message)-1] != '\n' {
```

Comment:
Hmmm, I'm not sure about this. The reason the `\n` is included is because sometimes it's not there -- specifically when there's an ultra-long line, it'll be broken up into `first part` ... `last part\n`. The idea is to allow the client to know that, and let it decide whether or not to format on multiple lines. What do you think?

Reply:
It seems to make the life of the consumer harder without much in return. If we have multiple lines, we should have a different way to flag that it's a continuation than forcing the client to introspect the message in a non-obvious way. The expectation of being able to just do the "${message}" sort of interpolation seems sane and useful. For example, how many times do we have in Go something like "something bad happened: %v". If we had to expect a newline or not at the end of that %v, we'd have to parse it every single time to standardize on one way or the other.

Reply:
Fair enough that it's annoying to fiddle with the `\n` on the client. However, I still want clients to be able to know when a message was truncated (didn't have a trailing `\n`). So I've added a `"truncated": true` to the API JSON object (only present in the rare cases that occurs).